lib/sycamore/sycamore/evaluation/subtasks.py

```diff
@@ -0,0 +1,139 @@
+        for key, val in filters.items():
+            hybrid_query_match.append({"match_phrase": {key: val}})
+        query_body["query"]["knn"]["embedding"]["filter"]["bool"]["must"] = hybrid_query_match
+        return query_body
+
+    def _get_subtasks(self, document: Document) -> Document:
+        terms = [
+            term for term in self._subtask_data["subtask_instructions"].keys() if term in document.text_representation
+        ]
+
+        for term in terms:
+            subtask = self._subtask_data["subtask_instructions"][term].format(**document.properties["subtask_filters"])
+            sub_doc = OpenSearchQuery()
+            sub_doc.index = self._index
+
+            subtask += "Return only the code " + term + " alongside the amount found and no additional information."
+            qn_embedding = self._embedder.generate_text_embedding(subtask)
+
+            sub_doc.query = {
+                "_source": {"excludes": ["embedding"]},
+                "size": self._os_config.get("size", 20),
+                "query": {
+                    "knn": {
+                        "embedding": {
+                            "vector": qn_embedding,
+                            "k": self._os_config.get("neural_search_k", 100),
+                            "filter": {
+                                "bool": {
+                                    "must": [
+                                        {"match": {"text_representation": subtask}},
+                                    ],
+                                },
+                            },
+                        }
+                    }
+                },
+            }
+
+            if "llm" in self._os_config:
+                sub_doc.params = {"search_pipeline": self._os_config["search_pipeline"]}
+                sub_doc["query"]["ext"] = {
+                    "generative_qa_parameters": {
+                        "llm_question": subtask,
+                        "context_size": self._os_config.get("context_window", 10),
+                        "llm_model": self._os_config.get("llm", "gpt-3.5-turbo"),
+                    }
+                }
+                if self._os_config.get("rerank", False):
+                    sub_doc["query"]["ext"]["rerank"] = {"query_context": {"query_text": subtask}}
+            if "filters" in document.properties:
+                sub_doc["query"] = self._add_filter(sub_doc["query"], document.properties["filters"])
+
+            document.elements.append(sub_doc)
+        return document
+
+    def _get_results(self, element: Element) -> Element:
+        return self._query_executor.query(element)
+
+    def execute(self, ds: DocSet) -> DocSet:
+        # Convert docset of all questions to a docset of formulas
+        formula_ds = ds.flat_map(self._get_formulas)
+        subtasks_ds = formula_ds.map(self._get_subtasks)
+        results_ds = subtasks_ds.map_elements(self._get_results)
+
+        return results_ds.take_all()
```

Comment:
this execute function says it returns a docset - results_ds is a docset... results_ds.take_all() is a list of documents. What should we be doing here?